contracts/contracts/strategies/CurveAMOStrategy.sol

```diff
@@ -0,0 +1,702 @@
+
+        // Ensure solvency of the vault
+        _solvencyAssert();
+
+        emit Withdrawal(address(hardAsset), address(lpToken), hardAssetAmount);
+    }
+
+    /**
+     * @dev Remove Curve pool LP tokens from the gauge and
+     * do a one-sided remove of hardAsset or OTOKEN from the Curve pool.
+     * @param _lpTokens The amount of Curve pool LP tokens to be removed from the gauge
+     * @param coinIndex The index of the coin to be removed from the Curve pool. 0 = hardAsset, 1 = OTOKEN.
+     * @return coinsRemoved The amount of hardAsset or OTOKEN removed from the Curve pool.
+     */
+    function _withdrawAndRemoveFromPool(uint256 _lpTokens, uint128 coinIndex)
+        internal
+        returns (uint256 coinsRemoved)
+    {
+        // Withdraw Curve pool LP tokens from Curve gauge
+        _lpWithdraw(_lpTokens);
+
+        // Convert Curve pool LP tokens to hardAsset value
+        uint256 valueInEth = _lpTokens.mulTruncate(
+            curvePool.get_virtual_price()
+        );
+
+        if (coinIndex == hardAssetCoinIndex) {
+            valueInEth = valueInEth.scaleBy(decimalsHardAsset, decimalsOToken);
+        }
+
+        // Apply slippage to hardAsset value
+        uint256 minAmount = valueInEth.mulTruncate(uint256(1e18) - maxSlippage);
+
+        // Remove just the hardAsset from the Curve pool
+        coinsRemoved = curvePool.remove_liquidity_one_coin(
+            _lpTokens,
+            int128(coinIndex),
+            minAmount,
+            address(this)
+        );
+    }
+
+    /**
+     * Checks that the protocol is solvent, protecting from a rogue Strategist / Guardian that can
+     * keep rebalancing the pool in both directions making the protocol lose a tiny amount of
+     * funds each time.
+     *
+     * Protocol must be at least SOLVENCY_THRESHOLD (99,8 %) backed in order for the rebalances to
+     * function.
+     */
+    function _solvencyAssert() internal view {
+        uint256 _totalVaultValue = IVault(vaultAddress).totalValue();
+        uint256 _totalOtokenSupply = oToken.totalSupply();
+
+        if (
+            _totalVaultValue.divPrecisely(_totalOtokenSupply) <
+            SOLVENCY_THRESHOLD
+        ) {
+            revert("Protocol insolvent");
+        }
+    }
+
+    /***************************************
+                Assets and Rewards
+    ****************************************/
+
+    /**
+     * @notice Collect accumulated CRV (and other) rewards and send to the Harvester.
+     */
+    function collectRewardTokens()
+        external
+        override
+        onlyHarvester
+        nonReentrant
+    {
+        // Collect CRV rewards from inflation
+        minter.mint(address(gauge));
+
+        // Collect extra gauge rewards (outside of CRV)
+        gauge.claim_rewards();
+
+        _collectRewardTokens();
+    }
+
+    function _lpWithdraw(uint256 _lpAmount) internal {
+        require(gauge.balanceOf(address(this)) >= _lpAmount, "Insufficient LP tokens");
```

Comment:
Nice. This is a good extra check that makes debugging txs easier.

Reply:
credits to @shahthepro! https://github.com/OriginProtocol/origin-dollar/pull/2436#discussion_r2007503037